File: ml/tree.py
"""Decision Tree classifier and regressor."""

from __future__ import annotations

import numpy as np


class DecisionTree:
    def __init__(self, max_depth: int = 10, min_samples_split: int = 2) -> None:
        self.max_depth: int = max_depth
        self.min_samples_split: int = min_samples_split
        self.tree: dict | None = None
        self.is_classification: bool | None = None
        self.n_classes: int = 0

    def fit(self, X: np.ndarray, y: np.ndarray) -> None:
        y_int = y.astype(int)
        self.is_classification = len(np.unique(y_int)) < len(y)
        self.n_classes = len(np.unique(y_int))
        self.tree = self._build_tree(X, y_int, depth=0)

    def _best_split(self, X: np.ndarray, y: np.ndarray) -> dict | None:
        best_gain = -1.0
        best_split: dict | None = None
        n_samples, n_features = X.shape

        for feat_idx in range(n_features):
            thresholds = np.unique(X[:, feat_idx])
            if len(thresholds) < 2:
                continue
            for t in thresholds[1:]:
                left_mask = X[:, feat_idx] <= t
                right_mask = ~left_mask
                if np.sum(left_mask) < self.min_samples_split:
                    continue
                gain = self._information_gain(y, left_mask, right_mask)
                if gain > best_gain:
                    best_gain = gain
                    best_split = {"feat_idx": feat_idx, "threshold": t}

        return best_split

    def _information_gain(self, y: np.ndarray, left_mask: np.ndarray, right_mask: np.ndarray) -> float:
        parent_entropy = self._entropy(y)
        n = len(y)
        n_l, n_r = np.sum(left_mask), np.sum(right_mask)
        if n_l == 0 or n_r == 0:
            return 0.0
        e_l = self._entropy(y[left_mask])
        e_r = self._entropy(y[right_mask])
        child_entropy = (n_l / n) * e_l + (n_r / n) * e_r
        return parent_entropy - child_entropy

    def _entropy(self, y: np.ndarray) -> float:
        if len(y) == 0:
            return 0.0
        probs = np.bincount(y.astype(int)) / len(y)
        probs = probs[probs > 0]
        return -np.sum(probs * np.log2(probs))

    def _build_tree(self, X: np.ndarray, y: np.ndarray, depth: int) -> dict:
        if depth >= self.max_depth or len(y) < 2 * self.min_samples_split:
            return self._leaf_value(y)

        split = self._best_split(X, y)
        if split is None:
            return self._leaf_value(y)

        left_mask = X[:, split["feat_idx"]] <= split["threshold"]
        right_mask = ~left_mask
        if np.sum(left_mask) == 0 or np.sum(right_mask) == 0:
            return self._leaf_value(y)

        return {
            "feat_idx": split["feat_idx"],
            "threshold": split["threshold"],
            "left": self._build_tree(X[left_mask], y[left_mask], depth + 1),
            "right": self._build_tree(X[right_mask], y[right_mask], depth + 1),
        }

    def _leaf_value(self, y: np.ndarray) -> dict:
        if len(y) == 0:
            return {"leaf": True, "value": 0}
        if self.is_classification:
            return {"leaf": True, "value": int(np.bincount(y.astype(int)).argmax())}
        return {"leaf": True, "value": float(np.mean(y))}

    def predict(self, X: np.ndarray) -> np.ndarray:
        return np.array([self._traverse(x, self.tree) for x in X])

    def _traverse(self, x: np.ndarray, node: dict) -> float | int:
        if node.get("leaf"):
            return node["value"]
        if x[node["feat_idx"]] <= node["threshold"]:
            return self._traverse(x, node["left"])
        return self._traverse(x, node["right"])

File: ml/tree.ts
type TreeNode = {
  feature: number;
  threshold: number;
  left: TreeNode;
  right: TreeNode;
} | {
  labels: number[];
};

function isLeaf(node: TreeNode): node is { labels: number[] } {
  return 'labels' in node;
}

export class DecisionTreeClassifier {
  max_depth: number;
  min_samples_split: number;
  n_features: number = 0;
  tree: TreeNode | null = null;

  constructor(max_depth = 10, min_samples_split = 2) {
    this.max_depth = max_depth;
    this.min_samples_split = min_samples_split;
  }

  gini(y: number[]): number {
    if (y.length === 0) return 0;
    const counts: Record<number, number> = {};
    for (const v of y) counts[v] = (counts[v] || 0) + 1;
    return 1 - Object.values(counts).reduce((s, c) => s + (c / y.length) ** 2, 0);
  }

  bestSplit(X: number[][], y: number[]): { feature: number; threshold: number; left_idx: number[]; right_idx: number[]; gain: number } | null {
    let best: { feature: number; threshold: number; left_idx: number[]; right_idx: number[]; gain: number } | null = null;
    const n = this.n_features || X[0].length;
    for (let f = 0; f < n; f++) {
      const vals = [...new Set(X.map(r => r[f]))].sort((a, b) => a - b);
      for (let i = 0; i < vals.length - 1; i++) {
        const t = (vals[i] + vals[i + 1]) / 2;
        const left_idx: number[] = [];
        const right_idx: number[] = [];
        for (let j = 0; j < X.length; j++) {
          if (X[j][f] <= t) left_idx.push(j); else right_idx.push(j);
        }
        if (left_idx.length === 0 || right_idx.length === 0) continue;
        const leftY = left_idx.map(idx => y[idx]);
        const rightY = right_idx.map(idx => y[idx]);
        const gain = this.gini(y) - (this.gini(leftY) * leftY.length / y.length) - (this.gini(rightY) * rightY.length / y.length);
        if (!best || gain > best.gain) {
          best = { feature: f, threshold: t, left_idx, right_idx, gain };
        }
      }
    }
    return best;
  }

  buildTree(X: number[][], y: number[], depth = 0): TreeNode | null {
    if (depth >= this.max_depth || y.length < this.min_samples_split) return { labels: y };
    const split = this.bestSplit(X, y);
    if (!split || split.gain <= 0) return { labels: y };

    const leftX = split.left_idx.map(i => X[i]);
    const rightX = split.right_idx.map(i => X[i]);
    const leftY = split.left_idx.map(i => y[i]);
    const rightY = split.right_idx.map(i => y[i]);

    const left = this.buildTree(leftX, leftY, depth + 1);
    const right = this.buildTree(rightX, rightY, depth + 1);

    if (!left || !right) return { labels: y };

    return { feature: split.feature, threshold: split.threshold, left, right };
  }

  fit(X: number[][], y: number[]): void {
    this.n_features = X[0].length;
    this.tree = this.buildTree(X, y);
  }

  predictOne(x: number[], node: TreeNode): number {
    if (isLeaf(node)) {
      const counts: Record<number, number> = {};
      for (const v of node.labels) counts[v] = (counts[v] || 0) + 1;
      return Number(Object.entries(counts).sort((a, b) => b[1] - a[1])[0]?.[0] ?? 0);
    }
    if (x[node.feature] <= node.threshold) return this.predictOne(x, node.left);
    return this.predictOne(x, node.right);
  }

  predict(X: number[][]): number[] {
    if (!this.tree) return [];
    return X.map(r => this.predictOne(r, this.tree!));
  }
}

export class DecisionTreeRegressor extends DecisionTreeClassifier {
  mse(y: number[]): number {
    if (y.length === 0) return 0;
    const m = y.reduce((s, v) => s + v, 0) / y.length;
    return y.reduce((s, v) => s + (v - m) ** 2, 0) / y.length;
  }

  predictOne(x: number[], node: TreeNode): number {
    if (isLeaf(node)) {
      return node.labels.reduce((s, v) => s + v, 0) / node.labels.length;
    }
    if (x[node.feature] <= node.threshold) return this.predictOne(x, node.left);
    return this.predictOne(x, node.right);
  }
}